rust/src/combinators/composition/sequential/mod.rs

```diff
@@ -0,0 +1,240 @@
+                        );
+
+                        // retrieve the last distance from d_mids, or bubble an error if d_mids is empty
+                        let d_mid = (d_mids.last())
+                            .ok_or_else(|| err!(FailedFunction, "out of queries"))?;
+
+                        // check that the query doesn't consume too much privacy
+                        if !measurement.check(&d_in, d_mid)? {
+                            return fallible!(FailedFunction, "insufficient budget for query");
+                        }
+
+                        // if the answer contains a queryable,
+                        // wrap it so that when the child gets a query it sends an AskPermission query to this parent queryable
+                        // it gives this sequential composition queryable (or any parent of this queryable)
+                        // a chance to deny the child permission to execute
+                        let child_id = d_mids.len() - 1;
+
+                        let mut sc_qbl = sc_qbl.clone();
+                        let wrap_logic = WrapFn::new_pre_hook(move || {
+                            sc_qbl.eval_internal(&AskPermission(child_id))
+                        });
+
+                        // evaluate the query and wrap the answer
+                        let answer = measurement.invoke_wrap(&arg, wrap_logic.as_map());
+
+                        // we've now consumed the last d_mid. This is our only state modification
+                        d_mids.pop();
+
+                        // done!
+                        return answer.map(Answer::External);
+                    }
+
+                    // if the query is internal (passed by the framework)
+                    if let Query::Internal(query) = query {
+                        // check if the query is from a child queryable who is asking for permission to execute
+                        if let Some(AskPermission(id)) = query.downcast_ref() {
+                            // deny permission if the sequential compositor has moved on
+                            if *id != d_mids.len() {
+                                return fallible!(
+                                    FailedFunction,
+                                    "sequential compositor has received a new query"
+                                );
+                            }
+                            // otherwise, return Ok to approve the change
+                            return Ok(Answer::internal(()));
+                        }
+                    }
+
+                    fallible!(FailedFunction, "unrecognized query!")
+                })
+            }
+        )),
+        input_metric,
+        output_measure,
+        PrivacyMap::new_fallible(move |d_in_p: &MI::Distance| {
+            if d_in_p.total_gt(&d_in)? {
+                fallible!(
+                    RelationDebug,
+                    "input distance must not be greater than d_in"
```

Comment:
In the context of this error message, it's a little confusing what `d_in` refers to. Though I'm not sure what to call it. Original input distance? Constructed input distance?

Reply:
I renamed `d_in_p` to `actual_d_in`. The actual `d_in` must be lte the `d_in` that the compositor is calibrated for.